### ISSUE TITLE
plugin can't handle volumes in docker-compose v2

### ISSUE THREAD

**Issue by holgerreif** [gondor/docker-volume-netshare#21]
We successfully used the plugin for nfs mounts with `docker volume create`, followed by `docker run -v`.
However with docker-compose we ran into a problem (http://jbu.io/2015/10/23/mount-nfs-efs-or-cifssamba-volumes-in-docker/#comment-2584967532).

As far as I understand we have the follwing problem: Given this `docker-compose.yml` file in a projectdir `test`:
```
version: '2'

volumes:
  data:
    driver: nfs

services:
  some_service:
    image: my_image
    volumes:
      - data:/mnt
```
docker-compose asks the plugin to create a volume `test_data` which in turn tries to mount something from a host `test_data` (which of course does not exist).

Solution would be to add a driver option to be used as follows
```
volumes:
  data:
    driver: nfs
      driver_opts:
        share: my.nfs.host:/path/to/share
```
Thus the plugin knows that the requested volume with name `test_data` refers to a share `my.nfs.host:/path/to/share` to be mounted under `/var/lib/docker-volumes/netshare/nfs/my.nfs.host/path/to/share`.  

This way we are even compatible with mixed `docker-compose` and pure `docker `client runs.

BTW: an in advance `docker volume create`d volume can be used the following way
```
volumes:
  data:
    external:
      name: my.nfs.host/path/to/share
```
So instead my proposed option `share: my.nfs.host:/path/to/share` you might implement `name: my.nfs.host/path/to/share`

**Comment by holgerreif** [comment]
Maybe you got my proposal wrong since you tagged it "Blocked - 3rd Party". This is something you could do with w/o any changes to docker-compose. Should I explain in more detail?

**Comment by gondor** [comment]
Yes, please ellaborate

**Comment by holgerreif** [comment]
Okay, let's do it. Actually I have two proposal: one workaround and the other is something to be implemented in docker-volume-netshare.

### Workaround: How to mount a nfs share with docker-volume-netshare in a docker-compose env 
[ I just write this down as an example on how to use it ]
First mount a share with 

`docker volume create -d nfs  --name my.nfs.host/path/to/share`

Then in your compose file you can use this share as follows
```
volumes:
  data:
    external:
      name: my.nfs.host/path/to/share
```
[ this is from scratch but I tested it before ]
disadvantage: you have to provide the share externally .
### usage of existing docker-compose file syntax
A volume in a [docker-compose file](https://docs.docker.com/compose/compose-file/#volume-configuration-reference) you can provide a `driver` and `driver_opts`. Th einteresting part is `driver_opts`. As _Those options are driver-dependent_ I'd expect, that docker-compose forwwards those otions to docker daemon and docker daemon in turn would forward it to the driver(like `--opt`of `docker volume`).

Thus I _think_ it is possible to specif the following as a docker-compose file:
```
volumes:
  data:
    driver: nfs
      driver_opts:
        share: my.nfs.host:/path/to/share
```
I'd expect, that option in last line would be handed over to volme plugin as key-value Pair `share=my.nfs.host:/path/to/share`. docker-volume-compose could use this option to identify, which host:/share combination should be mounted als volume `data:` instead of `data`. 

This compose file is a valif one (i.e. docker-compose will accept it), however it does not work since option `share` is not evaluated by volume plugin (this is your turn ;-)

These are only ideas, I hope you get the ghist and have the skills to find out, whether is could work along those lines!
(If not, just tell me what is unclear and I try to dig a bit deeper)

**Comment by gondor** [comment]
Ok totally makes sense.  Thanks for the extra info.  I'll play around with it sometime this weekend.  Currently finishing up BlueGreen support in my Depcon project. Will tackle this after.

**Comment by gondor** [comment]
@holgerreif I've tried everything. The driver opts seem to never make it into the plugin.  I have found other similar issues filed on docker-compose regarding driver_opts so I'm convinced this is a known bug.  

I've beefed up debug in each call we get from the daemon and am logging the current request and options coming in.  Options are always empty when using the driver_opts.  If you enable --verbose in the driver plugin for NFS you should see this too in the new 0.14 binary.  

If you agree, I'll flag this as blocked and we can monitor docker-compose for a fix.

**Comment by holgerreif** [comment]
Hmmm... looking throught the issues at [docker/compose](https://github.com/docker/compose) the most issues didn't really dealt with `driver_opts` but with a bug in the driver implementation itself. In my test the correct driver (docker-volume-netshare) was used, so it seems to be a different problem.
However one specific [issue #2899](https://github.com/docker/compose/issues/2899) seems to suggest, that some had working driver_opts with [Rex-ray](http://rexray.readthedocs.org/en/stable/) driver (problem was with type parsing, which is unreleated). I try to dig a bit into the code of the rexray volume driver on order to get a glue about it...

**Comment by gondor** [comment]
Please try on your stack. It could be versioning on my end.  The driver will print the incoming request from the docker daemon on every call in --verbose mode.  So if you figure out what in compose triggers the params coming through from the daemon, send it my way

**Comment by holgerreif** [comment]
I'm quite curious. Using `docker 1.10.3 9e83765` I do the follwoing:
```
docker volume create -d nfs --opt username=user --name host/share
```
I can't see any log entry like `Entering Create`. 
If I do 
```
docker run -it -v  host/share:/mnt --rm  some/image /bin/bash
```
the log entries `Entering Mount` and the like are visible, however the options are shown as empty `map[]`. 

I havn't tried it (will do tomorrow), but since I followed the lines alonge the cifs example 
```
$ docker volume create -d cifs --name cifshost/share --opt username=user ...
```
from [README.md](https://github.com/gondor/docker-volume-netshare/blob/master/README.md) I would expect to see the options somehow...

**Comment by gondor** [comment]
Are you running the 0.14 binary with --verbose ? Also, I didn't add the debug in CIFS, just the NFS route right now

**Comment by holgerreif** [comment]
I did:
```
$ egrep -v '^(#|$)' /etc/sysconfig/docker-volume-netshare
DKV_NETSHARE_OPTS="nfs -v 3 --verbose=true"
```
The other debug messages appear as expected:

```
time="2016-04-05T16:58:47+02:00" level=debug msg="Entering Mount: {host/share map[]}"
time="2016-04-05T16:59:08+02:00" level=debug msg="Entering Unmount: {host/share map[]}"
time="2016-04-05T17:03:12+02:00" level=debug msg="Entering Get: {host/share map[]}"
time="2016-04-05T17:03:12+02:00" level=debug msg="Entering REmove: name: host/share, options map[]"
time="2016-04-05T17:03:42+02:00" level=debug msg="Entering Get: {host/share map[]}"
```

Perhaps I'll retest with anothzer docker engine version...

**Comment by gondor** [comment]
Thats what I see too.  The map[] (^^ which shows empty), is where the driver_opts get added.  Using compose I don't see them being forward through. Glad your seeing similar results

**Comment by holgerreif** [comment]
I did setup a samba server on host `host`providing a share `share` accessable by user `user`with password `password`. The following worked:
```
docker-volume-netshare cifs -u username -p password &
docker volume create -d cifs --name host/share
docker run -it -v  host/share:/mnt --rm  some/image /bin/bash
```
However the following run into `mount error(13): Permission denied`
```
docker-volume-netshare cifs -u username &
docker volume create -d cifs --name host/share --opt password=password
docker run -it -v  host/share:/mnt --rm  some/image /bin/bash
```
It suggests, that the problem is with docker engine (again, 1.10.3).

This is even more surprising, since this is more or less taken from the example in [README.md](https://github.com/gondor/docker-volume-netshare/blob/master/README.md). Do yo know, which version of docker engine you used, when you tested this example?

Anyway, this looks like you should open an issue against docker project. I lokked through open issues w/o any success.

**Comment by gondor** [comment]
I haven't tested samba in a while to be honest.  It seems like more people use NFS so have had very little requests/issues related to it since I tested it last.  At the time I was using Docker 1.9.1 I believe.  Is there a quick/easy way to setup the Samba server for these tests?  I currently don't have one setup but have plenty of cloud/resources I can test with.   Just don't have  a lot of time to spend creating a server.

**Comment by holgerreif** [comment]
I did a downgrade from docker-1.10.3 to docker-1.9.1. What should I tell you? It worked:

### 1. With credentials provided at plugin start (works with 1.10.3 too)
```
$ docker-volume-netshare cifs -u username -p password &
[1] 37447
WARN[0000] Error: open /root/.netrc: no such file or directory
$ docker volume create -d cifs --name host/share
host/share
$ docker run -it -v  host/share:/mnt --rm alpine ash
INFO[0128] Mount: host/share, map[]
INFO[0128] Mounting CIFS volume //host/share on /var/lib/docker-volumes/netshare/cifs/host/share
/ # exit
INFO[0149] Unmounting volume //host/share from /var/lib/docker-volumes/netshare/cifs/host/share
$ docker volume rm //host/share
//host/share
```
### 2. Correct credential provided at create time - works as expected
```
$ docker-volume-netshare cifs  &
[1] 47512
WARN[0000] Error: open /root/.netrc: no such file or directory
$ docker volume create -d cifs --name host/share --opt username=username --opt password=password
host/share
root@tstsrdzsltc:~ # docker run -it -v  host/share:/mnt --rm alpine ash
INFO[0037] Mount: host/share, map[]
INFO[0037] Mounting CIFS volume //host/share on /var/lib/docker-volumes/netshare/cifs/host/share
/ # exit
INFO[0043] Unmounting volume //host/share from /var/lib/docker-volumes/netshare/cifs/host/share
$ docker volume rm  //host/share
//host/share
``` 
### 3. Wrong credentials provided at create time - fails as expected
```
$ docker volume create -d cifs --name host/share --opt username=username --opt password=pass
host/share
$ docker run -it -v  host/share:/mnt --rm alpine ash         INFO[0091] Mount: host/share, map[]
INFO[0091] Mounting CIFS volume //host/share on /var/lib/docker-volumes/netshare/cifs/host/share
2016/04/06 11:37:47 mount error(13): Permission denied
Refer to the mount.cifs(8) manual page (e.g. man mount.cifs)

INFO[0092] Unmounting volume //host/share from /var/lib/docker-volumes/netshare/cifs/host/share
2016/04/06 11:37:47 umount: /var/lib/docker-volumes/netshare/cifs/host/share: not mounted

Error response from daemon: Cannot start container 63a67a6d8baf7b14c8282f0758be7b7807ce2923bf4aa276bcd4c141ffc6f33d: Plugin Error: VolumeDriver.Mount, {"Mountpoint":"","Err":"exit status 32","Volumes":null,"Volume":null}
$ docker volume rm  //host/share                             //host/share
```
### 4. Finally just to test argument passing 
```
docker volume create -d nfs --name host/share --opt blah=blub
```
And the log file tells us:
```
time="2016-04-06T11:39:46.815205876+02:00" level=info msg="POST /v1.21/volumes/create"
time="2016-04-06T11:39:46+02:00" level=debug msg="Entering Create: name: host/share, options map[blah:blub]"
time="2016-04-06T11:39:46+02:00" level=debug msg="Create volume -> name: host/share, map[blah:blub]"
time="2016-04-06T11:39:46+02:00" level=debug msg="Entering Path: {host/share map[]}"
time="2016-04-06T11:39:46+02:00" level=debug msg="Path for host/share is at /var/lib/docker-volumes/netshare/nfs/host/share"
```
So it looks like a **bug in docker engine 1.10.3**!

**Comment by holgerreif** [comment]
> So it looks like a bug in docker engine 1.10.3!

After upgrade to 1.10.3 again I found an interesting fact. Running 
```
$ docker volume create -d nfs --name host/share2 --opt blah=blub host/share2
host/share2
$ docker volume ls
DRIVER              VOLUME NAME
cifs                /var/lib/docker-volumes/netshare/cifs/host/share
```
The following output was logged
```
time="2016-04-06T16:11:19+02:00" level=debug msg="Entering Get: {host/share2 map[]}"
time="2016-04-06T16:11:19+02:00" level=debug msg="Get for host/share2 is at /var/lib/docker-volumes/netshare/nfs/host/share2"
time="2016-04-06T16:11:38+02:00" level=debug msg="Entering List: { map[]}"
```
Interestingly. Instead of calling `Create` from the plugin it is called with `Get`.

Even more interesting: The first `docker volume create` after plugin restart is not even forwarded to the plugin! Only after `volume rm` the next `volume create` is forwared as `Get`.

I think, you should open an issue at [docker/docker](https://github.com/docker/docker). If you want I can do it as well, however I might need some support onece a few more technical question arise.

**Comment by holgerreif** [comment]
From [issue (comment)](https://github.com/docker/docker/issues/21857#issuecomment-207081492) it seems the first `Get` Call to the plugin returns something. As non-Go-skilled guy I'm unable to decide looking at the code what the plugin actually returns, but it sounds realistic.

@gondor Could you look at this?

**Comment by holgerreif** [comment]
Okay, I _think_ I got it:`Get()` in `nfs.go` always returns something! It returns, where a volume _would have been mounted_ (the same as `Path()`) independent whether this volume already exists.

Solution: you have to keep track of created mounts!
1. add a remove function for mountManager which deletes the rnty from m.mounts
2. nfsDriver.Remove function should call mountManager.Remove
3. nfsDriver.Get should check, whether mountManager knows about the volume (e.g. call GetVolumes and hcek whether this specific volume is in the returned map or implement a mountManager.GetVolume (VolName) function.

Once solved we could come back to the original problem ;-)

**Comment by gondor** [comment]
Try this version below.  I added the logic.

I've noticed however that in some cases where volumes fail to mount (Mount 32 error) you need to do the trailing /.  This has changed in 1.10.3 of docker (didn't need it before).  For example:

```
docker volume create -d nfs --name nfs.server.com/mnt/share/
docker run -it --rm -v nfs.server.com/mnt/share/:/data ubuntu bash
```

[docker-volume-netshare-15dev.zip](https://github.com/gondor/docker-volume-netshare/files/209693/docker-volume-netshare-15dev.zip)

**Comment by swoopster-sb** [comment]
`docker volume create -d nfs --name nfs.server.com/mnt/share/`
unfortunately gives
`Error response from daemon: unable to find a node that satisfies node==nfs.server.com`
when running on a docker swarm.

I guess the best solution would be to add a parameter to pass the "nfs.server/share" to the driver as option (not in the name of the share)

**Comment by holgerreif** [comment]
I played a little bit with your version. From looking at the --verbose output I I think it works as expected. Have not tested docker-compose yet (I'll finde some time over weekend).

One thing: instead of
```
Get: mount found for host/share -> /var/lib/docker-volumes/netshare/nfs/host/share
```
the log should read
```
Get: volume description found for host/share -> /var/lib/docker-volumes/netshare/nfs/host/share
```
Actually, according to the code it is in the `mount[]`, however technically speaking it's not mounted yet ;-)

However cifs is not working, probably because you have not changed it's implementation. However following the same logis as nfs it should work.

**Comment by holgerreif** [comment]
@swoopster-sb I havn't actually tested with swarm, however the proposed parameter `share` for `docker-compose file` in my [previous comment](https://github.com/gondor/docker-volume-netshare/issues/21#issuecomment-204806589) would perfectly solve your problem as well ;-)

**Comment by gondor** [comment]
I ended up doing a major refactor to simplify the same boiler plate between each driver.  This binary and associated branch I checked into reflects this.  I also added the share option.

For example:
```
docker volume create -d nfs --name myvol -o share=myserver:/share
docker run -it --rm -v myvol:/data ubuntu bash
```
This also works too (but noticed on my install daemon 1.10.3) wants the trailing /
```
docker run -it --rm -v nfs.server.orgmnt/share/:/data --volume-driver=nfs ubuntu bash
```

CIFS and EFS inherited the NFS changes but I didn't test.. So test NFS then CIFS to compare.  They all support the "share" option too

This refactor and binary also added the output on start to cover Issue #25 so can you close that is you are playing with this.

[docker-volume-netshare-16.zip](https://github.com/gondor/docker-volume-netshare/files/211134/docker-volume-netshare-16.zip)

**Comment by holgerreif** [comment]
Refactoring was an idea I had too. I'm glad you did it.

My test show, that `share` option works great. 
I tested mit driver nfs both `docker volume create`and `docker run --volume-driver`. 

BTW: For me it worked both with and w/o the trailing slash.

And -- keep your seatbelts fastened ;-) -- `docker-compose` can work now with your plugin as well!

I was able to get the follwoing `docker-compose.yml` to work.
```
version: '2'

volumes:
  data:
    driver: nfs
    driver_opts:
      share: host:/share

services:
  hello:
    image: alpine
    volumes:
      - data:/mnt
    command: ls /mnt
```
cifs however got broken, neither example [from my comment](https://github.com/gondor/docker-volume-netshare/issues/21#issuecomment-206388228) does not work anymore, besides opts are handed passed to . 

I'll treat this issue (usage in docker-compose) as solved (and leave it up to you to close). After digging a bit more I'll open a separate issue for broken cifs functionality.

**Comment by gondor** [comment]
Awesome, thanks for testing this.  I'll see if I can figure out how to test CIFS.  I'm all OSX and Linux so typically use NFS based variants.  Let's create a separate ticket for CIFS

**Comment by holgerreif** [comment]
@swoopster-sb Meanwhile I understood why you had probs with swarm: `host:/share` is the syntax to get a local share on a specific node in cluster.
Could you re-test within your swarm setup and post the result here as well as the outup of `docker volume ls`. Should serve as a reference!

**Comment by swoopster-sb** [comment]
```
docker volume create --driver=nfs --name gv0 -o share=nfs.server.com:/gv0
docker volume ls
```
gives
```
DRIVER              VOLUME NAME
nfs                 gv0
nfs                 gv0
```
not sure if it is a good sign that it shows up twice (currently running 2 nodes in swarm, maybe its because of that!?)

`docker run -it --rm -v gv0:/data  alpine ls /data` works fine.

Also seems to be working in docker-compose with

```
volumes:
 gv0:
  driver: nfs
  driver_opts:
   share: "nfs.server.net:/gv0"
```

**Comment by holgerreif** [comment]
> not sure if it is a good sign that it shows up twice

@swoopster-sb  Actually the plugin on both nodes creates a volume. For listing it reports the same volume on both nodes since they do not know from each other (in contrats to a storage cluster driver). What is the expected listing? What is logged from the plugin wenn started in `--verbose`mode?

**Comment by gondor** [comment]
This is actually expected.  Swarm will call out to all nodes on "ls, create, etc".  Even the test cases around swarm and volume show this:

https://github.com/docker/swarm/blob/master/test/integration/api/volume.bats

**Comment by swoopster-sb** [comment]
ok, so i think it works perfectly as expected! Nice.
